waspls/src/Wasp/LSP/Handlers.hs

```diff
@@ -67,50 +74,37 @@ didSaveHandler =
 -- file in "Wasp.LSP.State.State".
 diagnoseWaspFile :: LSP.Uri -> ServerM ()
 diagnoseWaspFile uri = do
-  src <- readVFSFile uri
-  let appSpecOrError = Wasp.Analyzer.analyze $ T.unpack src
-  diagnostics <- case appSpecOrError of
-    -- Valid wasp file, send no diagnostics
-    Right _ -> return $ LSP.List []
-    -- Report the error (for now, just one error per analyze is possible)
-    Left err ->
-      return $
-        LSP.List
-          [ waspErrorToLspDiagnostic err
-          ]
+  updateState uri
+  currentDiagnostics <- gets (^. diagnostics)
   liftLSP $
     LSP.sendNotification LSP.STextDocumentPublishDiagnostics $
-      LSP.PublishDiagnosticsParams uri Nothing diagnostics
-  where
-    waspErrorToLspDiagnostic :: WE.AnalyzeError -> LSP.Diagnostic
-    waspErrorToLspDiagnostic err =
-      let errSrc = case err of
-            WE.ParseError _ -> "parse"
-            WE.TypeError _ -> "typecheck"
-            WE.EvaluationError _ -> "evaluate"
-          (errMsg, errCtx) = WE.getErrorMessageAndCtx err
-       in LSP.Diagnostic
-            { _range = waspCtxToLspRange errCtx,
-              _severity = Nothing,
-              _code = Nothing,
-              _source = Just errSrc,
-              _message = T.pack errMsg,
-              _tags = Nothing,
-              _relatedInformation = Nothing
-            }
+      LSP.PublishDiagnosticsParams uri Nothing (LSP.List currentDiagnostics)
 
-    waspCtxToLspRange :: Ctx -> LSP.Range
-    waspCtxToLspRange (Ctx region) =
-      LSP.Range
-        { _start = waspSourcePositionToLspPosition (getRgnStart region),
-          -- Increment end character by 1: Wasp uses an inclusive convention for
-          -- the end position, but LSP considers end position to not be part of
-          -- the range.
-          _end = waspSourcePositionToLspPosition (getRgnEnd region) & LSP.character +~ (1 :: LSP.UInt)
-        }
-
-    waspSourcePositionToLspPosition (WE.SourcePosition l c) =
-      LSP.Position (fromIntegral $ l - 1) (fromIntegral $ c - 1)
+updateState :: LSP.Uri -> ServerM ()
+updateState uri = do
+  src <- readVFSFile uri
+  let srcString = T.unpack src
+  modify (sourceString .~ srcString)
+  let concreteParse = parseCST $ L.lex srcString
+  -- Put CST in state
+  modify (cst ?~ snd concreteParse)
+  if not $ null $ fst concreteParse
+    then do
+      -- Errors found during concrete parsing. Replace diagnostics with new parse
+      -- diagnostics.
+      newDiagnostics <- mapM (concreteParseErrorToDiagnostic srcString) $ fst concreteParse
+      modify (diagnostics .~ newDiagnostics)
+    else do
+      -- No concrete parse errors, run full analyzer!
+      let analyzeResult = analyze srcString
+      case analyzeResult of
+        Right _ -> do
+          modify (diagnostics .~ [])
+        Left err -> do
+          let newDiagnostics =
+                [ waspErrorToDiagnostic err
+                ]
+          modify (diagnostics .~ newDiagnostics)
```

Comment:
Let's improve `updateState` name to smth nicer. What does it do really?

Also, can we make code a bit simpler / better organized? It is a bit hard to read at the moment.
For example `modify (sourceString .~ srcString)` -> hmm what does that do?

Reply:
Renamed to `analyzeWaspFile`: it's responsible for taking source text and parsing it to produce CST and possibly error messages.

I've tried to reorganize things to make them more readable. And yeah, those modify lines are a little hard to read, but it's setting the `sourceString` field in the `ServerState` to `srcString`.